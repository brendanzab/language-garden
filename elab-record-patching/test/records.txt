let F := {
  A : Type;
  B : Type;
  f : A -> B;
};

let patch-1   : F [ B := A ] -> F                                       := fun x => x;
let patch-2   (A : Type) : F [ A := A; B := A ] -> F                    := fun x => x;
let patch-3   (A : Type) : F [ A := A; B := A; f := fun x => x ] -> F   := fun x => x;
let patch-3b  (A : Type) : F [ B := A; f := fun x => x; A := A; ] -> F  := fun x => x;
let patch-4   (A : Type) : F [ A := A; B := A ] -> F [ B := A ]         := fun x => x;
let patch-5   (A : Type) : F [ A := A; B := A ] -> F [ A := A ]         := fun x => x;
let patch-6   (C : Type) : F [ A := C; B := C ] -> F [ B := C ]         := fun x => x;

let coerce-missing-1 (C : Type) : F [ A := C; B := C ]                   := { f := fun x => x };
let coerce-missing-2 (C : Type) : F [ A := C; B := C; f := fun x => x ]  := ({} : {});
let coerce-missing-3 (C : Type) : F [ A := C; B := C ]                   := { A := C; f := fun x => x };

let record-lit-coerce-1 (B : Type) : { A : Type [= B]; a : B } -> { A : Type; a : A } :=
  fun r => r;
let record-lit-coerce-2 (B : Type) (b : B) : { A : Type; a : A } :=
  { A := B; a := b } : { A : Type; a : B };

let coerce-missing-patched-fields-1 (A : Type) (B : Type) : { f : A -> B } -> F [ A := A; B := B ] :=
    fun r => r;
let coerce-missing-patched-fields-2 (A : Type) (B : Type) : { A : Type; f : A -> B } -> F [ B := B ] :=
    fun r => r;

Type
