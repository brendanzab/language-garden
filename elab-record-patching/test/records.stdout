<input> : Type :=
  let F := { A : Type; B : Type; f : A -> B };
  let patch-1 : { A : Type; B : Type [= A]; f : A -> A } -> F :=
    fun x := { A := x.A; B := #sing-elim x.B x.A; f := x.f };
  let patch-2 :
        fun (A : Type)
            { A : Type [= A]; B : Type [= #sing-elim A A]; f : A -> A } -> F
  := fun A x := { A := #sing-elim x.A A; B := #sing-elim x.B A; f := x.f };
  let patch-3 :
        fun (A : Type)
            {
              A : Type [= A];
              B : Type [= #sing-elim A A];
              f : (A -> A) [= fun x := x]
            } -> F
  :=
    fun A x :=
      {
        A := #sing-elim x.A A;
        B := #sing-elim x.B A;
        f := #sing-elim x.f (fun x := x)
      };
  let patch-3b :
        fun (A : Type)
            {
              A : Type [= A];
              B : Type [= #sing-elim A A];
              f : (A -> A) [= fun x := x]
            } -> F
  :=
    fun A x :=
      {
        A := #sing-elim x.A A;
        B := #sing-elim x.B A;
        f := #sing-elim x.f (fun x := x)
      };
  let patch-4 :
        fun (A : Type)
            { A : Type [= A]; B : Type [= #sing-elim A A]; f : A -> A } ->
            { A : Type; B : Type [= A]; f : A -> A }
  := fun A x := { A := #sing-elim x.A A; B := x.B; f := x.f };
  let patch-5 :
        fun (A : Type)
            { A : Type [= A]; B : Type [= #sing-elim A A]; f : A -> A } ->
            { A : Type [= A]; B : Type; f : A -> B }
  := fun A x := { A := x.A; B := #sing-elim x.B A; f := x.f };
  let patch-6 :
        fun (C : Type) { A : Type [= C]; B : Type [= C]; f : C -> C } ->
            { A : Type; B : Type [= C]; f : A -> C }
  := fun C x := { A := #sing-elim x.A C; B := x.B; f := x.f };
  let coerce-missing-1 :
        fun (C : Type) -> { A : Type [= C]; B : Type [= C]; f : C -> C }
  := fun C := { A := #sing-intro; B := #sing-intro; f := fun x := x };
  let coerce-missing-2 :
        fun (C : Type)
            ->
            { A : Type [= C]; B : Type [= C]; f : (C -> C) [= fun x := x] }
  := fun C := { A := #sing-intro; B := #sing-intro; f := #sing-intro };
  let coerce-missing-3 :
        fun (C : Type) -> { A : Type [= C]; B : Type [= C]; f : C -> C }
  := fun C := { A := #sing-intro; B := #sing-intro; f := fun x := x };
  let coerce-missing-patched-fields-1 :
        fun (A : Type) (B : Type) { f : A -> B } ->
            { A : Type [= A]; B : Type [= B]; f : A -> B }
  := fun A B r := { A := #sing-intro; B := #sing-intro; f := r.f };
  let coerce-missing-patched-fields-2 :
        fun (A : Type) (B : Type) { A : Type; f : A -> B } ->
            { A : Type; B : Type [= B]; f : A -> B }
  := fun A B r := { A := r.A; B := #sing-intro; f := r.f }; Type
